lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -203,9 +203,10 @@ void ExecutionEngine::optimizeFunction(CompilationMode mode, Function *F) {
   }
 }
 
-void ExecutionEngine::compile(CompilationMode mode, Function *F,
-                              const Context &ctx) {
+void ExecutionEngine::compile(CompilationMode mode, Function *F, Context &ctx) {
   optimizeFunction(mode, F);
+  // Make sure that the context has backing tensors for all placeholders.
+  ctx.allocate(M_.getPlaceholders());
```

Comment:
Do we have some philosophy on correct usage for allocating Placeholders? Adding this call here means we could/will have inconsistent usage, where some users will allocate Placeholders prior to calling `compile()`, and others will just rely on this `allocate()` call here.

Reply:
At the moment we are passing the context to the compile method. This means that the compile method has to remember the Placeholder buffers. The next step in this rewrite would be to move the context from the compile method to the execute method. This will allow multithreaded execution of compiled methods because the placeholders addresses won't be baked into the compiled function. So, initialization of the context has to happen in the test somewhere after lowering, but before execution. @jfix71

Reply:
Moving the context parameter to the backends would require some work because at the moment the CPU backend assumes that it can use the pointers as constants. So, we'll need to adopt the AOT style compilation. To summarize, I'd like to move the line that I added here out side the compile method, but it will probably take a few weeks. @jfix71